lib/Interface/cpp_interface.cpp

```diff
@@ -613,4 +640,47 @@ void CNFClearMemory()
 {
   Cnf_ClearMemory();
 }
+
+Cpp_interface::SolverFrame::SolverFrame(
+    std::unordered_map<std::string, Function>* global_function_context)
+    : _global_function_context(global_function_context)
+{
+}
+
+// When we destroy a solver frame, we need to make sure that all of the scoped
+// functions in the global function context are also correctly removed.
+//
+// This ensures that the reference counting for any symbols used in the
+// function declarations are correctly decremented.
+Cpp_interface::SolverFrame::~SolverFrame()
+{
+  // Iterate on the function names in our current scope
+  for (vector<std::string>::const_iterator scoped_function_name =
+           getFunctions().begin();
+       scoped_function_name != getFunctions().end(); ++scoped_function_name)
+  {
+    // Find this function in the global context
+    std::unordered_map<std::string, Function>::iterator function_to_erase =
```

Comment:
You could make make this:
``
    std::unordered_map<const std::string, Function>::const_iterator function_to_erase =
``
or
```
const auto function_to_erase =
```
if you wanted to.